### PR TITLE
sketchy/rudimentary decoder for CrAU updates

### PR DESCRIPTION
**Pull request by wrl**
this PR sketches out a reader for the CrAU update format. there is an outer binary "wrapper" and then an inner protobuf. the documentation is wrong in several places (`update_metadata.proto` states that the manifest is bzip2 compressed, and it... just isn't), so approach any docs with heavy skepticism and don't be surprised if you have to dive into the update_engine c++.

the relevant file is `update-format-crau/src/bin/test.rs`. this loads an update delta file and prints out the signature blobs. there can be several of them – i don't know exactly what is being signed.

next steps are extracting the flow in `test.rs` and turning it into more of a library interface.

ref https://github.com/flatcar/Flatcar/issues/1025

### CODE REVIEW

**Comment by dongsupark** [comment]
Hmm, where is the input file `update_metadata.proto`?
I see only its generated file `update-format-crau/src/generated/update_metadata.rs`.

**Comment by dongsupark** [comment]
Ah, found [it](https://github.com/flatcar/update_engine/blob/c6f566d47d8949632f7f43871eb8d5c625af3209/src/update_engine/update_metadata.proto).

**Comment by pothos** [comment]
I've added generating the dev-key- signed sysext payloads here: https://github.com/flatcar/scripts/pull/919
You can download a test file from https://bincache.flatcar-linux.net/images/amd64/9999.9.9+kai-sign-test-oem-sysext/flatcar_test_update-oem-azure.gz

**Comment by dongsupark** [comment]
@pothos Thanks for the test data.
I have tested it, and it seems to work.
Added a README to describe how to test.

File: update-format-crau/src/bin/test.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+fn read_delta_update_header(f: &mut dyn Read) -> Result<DeltaUpdateFileHeader, Box<dyn Error>> {
+    let mut header = DeltaUpdateFileHeader {
+        magic: [0; 4],
+        file_format_version: 0,
+        manifest_size: 0,
+    };
+
+    f.read_exact(&mut header.magic)?;
+    if &header.magic != b"CrAU" {
```
Comment (dongsupark) [inline update-format-crau/src/bin/test.rs:33]: (Github simply ignored my last comment, so am rewriting)

Ideally let's define `b"CrAU"` as const to be reused elsewhere.

File: update-format-crau/src/bin/test.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+use std::io::{Read, Seek, SeekFrom};
```
[the remark this answers is not in the document]
Comment (wrl) [reply update-format-crau/src/bin/test.rs:1]: imo once the code has been converted into a library then that file can go away and a separate test bin that uses the library would replace it

File: update-format-crau/src/bin/test.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    manifest_size: u64,
+}
+
+impl DeltaUpdateFileHeader {
+    #[inline]
+    fn translate_offset(&self, offset: u64) -> u64 {
+        DELTA_UPDATE_HEADER_SIZE + self.manifest_size + offset
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (wrl) [reply update-format-crau/src/bin/test.rs:24]: honestly i don't even like having those fields in the DeltaUpdateFileHeader. magic and file version are only used for verification/validation, they're not useful past the initial load. I'd rather remove them from the header struct.

the header size variable... eh. if we had to support multiple header versions (or if there even *were* other header versions to support) then I'd see the point, but imo it doesn't clarify anything that some comments about the static calculation wouldn't do instead.

File: update-format-crau/src/bin/test.rs (hunk 1 above)
Comment (wrl) [reply update-format-crau/src/bin/test.rs:33]: sure, no issues. don't think it needs to be a pub export from the update-format-crau subcrate though.

File: update-format-crau/src/bin/test.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dongsupark) [reply update-format-crau/src/bin/test.rs:1]: ok

File: update-format-crau/src/bin/test.rs (hunk 1 above)
Comment (dongsupark) [reply update-format-crau/src/bin/test.rs]: Done, defined DELTA_UPDATE_FILE_MAGIC.

**Pull request merged**